### PR TITLE
refactor preprocessing stage

### PR DESCRIPTION
**Pull request by chenshanpei**
there are a few benifits from code refactor:

1.  code framwork becomes clear, three seperate files to three stages of preprocessing.
2.  put all data collection work (dirty work) to collection stage;
3.  get rid of fake.c
4.  speed up extraction stage (from 7s to 1s)

### CODE REVIEW

File: sched_boundary/sched_boundary.py [hunk 1]
```diff
@@ -117,50 +115,34 @@ def register_cbs(self):
...
-            signature,
-            decl.location.line - 1,
-            decl.location.column - 1,
-            decl.function.end.line - 1,
-            decl.function.end.column - 1])
-
-    def function_define(self, decl, _):
-        # only func definition will trigger PLUGIN_FINISH_PARSE_FUNCTION
+    def function_location(self):
         loc = gcc.get_location()
```
Comment (ampresent) [inline sched_boundary/sched_boundary.py:119]: I suspect whether this works? At final_work, `gcc.get_location` always returns the same code line despite where they're defined.

Comment (ampresent) [inline sched_boundary/sched_boundary.py:119]: I suspect whether this works? After moving it to final_work, `gcc.get_location` always returns the same code line despite where they're defined.

File: sched_boundary/sched_boundary.py [hunk 2]
```diff
@@ -144,38 +144,27 @@ def function_location(self):
...
+        if src_file == self.fake: return
 
-        if decl.name in self.config['global_var']['force_private']:
-            return
+        with open(tmp_config) as f:
+            self.meta = json.load(f)
 
-        # share public (non-static) variables by default
-        if decl.public or decl.name in self.config['global_var']['extra_public']:
-            if anonymous_type_var(decl):
```
Comment (ampresent) [inline sched_boundary/sched_boundary.py:186]: This is removed because ?

**Comment by ampresent** [comment]
You seem to miss pushing `extract.py`?

File: src/Makefile.plugsched [hunk 3]
```diff
@@ -22,5 +22,9 @@ collect:
 		CFLAGS_MODULE="$(GCC_PLUGIN_FLAGS)collect" $(vmlinux-deps)
 analyze:
 	python3 $(plugsched_tmpdir)/process.py ./vmlinux $(plugsched_tmpdir) $(plugsched_modpath)
-extract:
-	$(MAKE) CFLAGS_KERNEL="$(GCC_PLUGIN_FLAGS)extract" $(objs)
+
+extract: $(objs)
+
+%.extract:
```
Comment (ampresent) [inline src/Makefile.plugsched:28]: The target and prerequisite may be simplified. Refer to https://www.gnu.org/software/make/manual/html_node/Pattern-Rules.html

File: cli.py [hunk 4]
```diff
@@ -91,7 +91,7 @@ def __init__(self, work_dir, vmlinux, makefile, sym_vers):
         self.mod_files = self.config['mod_files']
         self.mod_srcs = [f for f in self.mod_files if f.endswith('.c')]
         self.mod_hdrs = [f for f in self.mod_files if f.endswith('.h')]
-        self.mod_objs = [f[:-2]+'.o' for f in self.mod_srcs] + ['kernel/sched/fake.o']
+        self.mod_objs = [f+'.extract' for f in self.mod_files]
```
Comment (ampresent) [inline cli.py:94]: It seems `fake.c` is not needed anymore. We can delete this file in this patch.

Comment (chenshanpei) [reply cli.py:93]: the last patch will remove fake.c.

File: src/Makefile.plugsched (hunk 3 above)
Comment (chenshanpei) [reply src/Makefile.plugsched]: _%.extract: %
          python3 $(plugsched_tmpdir)/extract.py $^ $(plugsched_tmpdir) $(plugsched_modpath)_

works for me.

**Pull request merged**